src/webgl/p5.RendererGL.Immediate.js

```diff
@@ -115,211 +92,136 @@ p5.RendererGL.prototype.endShape = function(
+          for (i = 0; i < im.vertices.length - 2; i += 1) {
+            im.edges.push([i, i + 1], [i, i + 2]);
+          }
+          im.edges.push([im.vertices.length - 2, im.vertices.length - 1]);
+          break;
+
+        default:
+          for (i = 0; i < im.vertices.length - 1; i++) {
+            im.edges.push([i, i + 1]);
+          }
+          if (this.drawMode !== constants.TEXTURE) {
+            if (mode === constants.CLOSE) {
+              im.edges.push([im.vertices.length - 1, 0]);
+            }
+          }
+          break;
+      }
+
+      im._edgesToVertices();
+    }
 
-  // initialize the fill shader's 'aVertexColor' buffer
-  if (
-    this.drawMode === constants.FILL &&
-    this.curFillShader.attributes.aVertexColor
-  ) {
-    this._bindBuffer(
-      this.immediateMode.colorBuffer,
-      gl.ARRAY_BUFFER,
-      this.immediateMode.vertexColors,
-      Float32Array,
-      gl.DYNAMIC_DRAW
-    );
-
-    this.curFillShader.enableAttrib(
-      this.curFillShader.attributes.aVertexColor.location,
-      4,
-      gl.FLOAT,
-      false,
-      0,
-      0
-    );
-  }
+    this._setDefaultCamera();
 
-  // initialize the fill shader's 'aTexCoord' buffer
-  if (
-    this.drawMode === constants.TEXTURE &&
-    this.curFillShader.attributes.aTexCoord
-  ) {
-    //texture coordinate Attribute
-    this._bindBuffer(
-      this.immediateMode.uvBuffer,
-      gl.ARRAY_BUFFER,
-      this.immediateMode.uvCoords,
-      Float32Array,
-      gl.DYNAMIC_DRAW
-    );
-
-    this.curFillShader.enableAttrib(
-      this.curFillShader.attributes.aTexCoord.location,
-      2,
-      gl.FLOAT,
-      false,
-      0,
-      0
-    );
-  }
+    this.createBuffers('immediate', im);
+    var geometry = this.gHash['immediate'];
 
-  //if (true || mode) {
-  if (this.drawMode === constants.FILL || this.drawMode === constants.TEXTURE) {
-    switch (this.immediateMode.shapeMode) {
-      case constants.LINE_STRIP:
-      case constants.LINES:
-      case constants.TRIANGLES:
-        this.immediateMode.shapeMode = constants.TRIANGLE_FAN;
-        break;
+    // select the stroke shader
+    var stroke = this.curStrokeShader;
+    if (!stroke || !stroke.isStrokeShader()) {
+      stroke = this._getLineShader();
     }
-  } else {
-    switch (this.immediateMode.shapeMode) {
-      case constants.LINE_STRIP:
-      case constants.LINES:
-        this.immediateMode.shapeMode = constants.LINE_LOOP;
-        break;
+    // render the fill
+    this._renderStroke(geometry, stroke);
+
+    if (drawMode) {
+      //select the fill shader
+      var fill = this.curFillShader;
+      if (this._enableNormal) {
+        fill = this._getNormalShader();
+      } else if (this._enableLighting) {
+        if (!fill || !fill.isLightShader()) {
+          fill = this._getImmediateLightShader();
+        }
+      } else {
+        if (!fill /*|| !fill.isColorShader()*/) {
+          fill = this._getImmediateFlatShader();
+        }
+      }
```

Comment:
because in immediate mode there's no such thing as a 'texture shader' - it's just based on the `isTexture` uniform. this behavior hasn't changed - it's how it works now.

Reply:
i never intended to unify the shaders completely (although the lighting calculation did get unified), simply because the requirements of the immediate and retained mode shaders are quite different. the main unification here is with the handling of the geometry buffers.

Reply:
okay. this is where having an issue filed that clearly states a feature or enhancement before a PR is submitted is helpful. please do this in the future!

Reply:
well, to be fair, i _did_ list several pre-existing issues that the original PR addressed.

Reply:
yes, but a refactor of this scale was not necessary to fix the individual issues. an issue to document enhancing the architecture would have been helpful.